File: src/main/java/day1java/example3/Outcome.java
package day1java.example3;

import java.util.function.Consumer;
import java.util.function.Function;
import java.util.function.Supplier;

public interface Outcome<ERR, T> {

    default <U> Outcome<ERR, U> map(Function<T, U> f) {
        if (this instanceof Success) {
            return new Success(f.apply(((Success<T>) this).value));
        } else if (this instanceof Failure) {
            return ((Failure) this);
        } else {
            throw new RuntimeException("Unknown type " + this.getClass().getSimpleName());
        }
    }


    default <ERR2> Outcome<ERR2, T> mapFailure(Function<ERR, ERR2> f) {
        if (this instanceof Success) {
            return ((Success) this);
        } else if (this instanceof Failure) {
            return new Failure(f.apply(((Failure<ERR>) this).error));
        } else {
            throw new RuntimeException("Unknown type " + this.getClass().getSimpleName());
        }
    }


    class Success<T> implements Outcome<Void, T> {
        public final T value;

        public Success(T value) {
            this.value = value;
        }
    }

    class Failure<ERR> implements Outcome<ERR, Void> {
        public final ERR error;

        public Failure(ERR error) {
            this.error = error;
        }
    }

    static <T> Outcome<Throwable, T> tryThis(Supplier<T> block) {
        try {
            return new Success(block.get());
        } catch (Throwable e) {
            return new Failure(e);
        }
    }


    default <U> Outcome<ERR, U> flatMap(Function<T, Outcome<ERR, U>> f) {
        if (this instanceof Success) {
            return f.apply(((Success<T>) this).value);
        } else if (this instanceof Failure) {
            return ((Failure) this);
        } else {
            throw new RuntimeException("Unknown type " + this.getClass().getSimpleName());
        }
    }


    default T onFailure(Consumer<ERR> block) {
        if (this instanceof Success) {
            return ((Success<T>) this).value;
        } else if (this instanceof Failure) {
            block.accept(((Failure<ERR>) this).error);
            throw new RuntimeException("Unmanged failure");
        } else {
            throw new RuntimeException("Unknown type " + this.getClass().getSimpleName());
        }
    }

}

File: src/main/kotlin/day1/example3/Outcome.kt
package day1.example3


sealed class  Outcome<out E, out T: Any> {

    fun <U: Any> map(f: (T) -> U): Outcome<E, U> =
        when (this){
            is Success -> Success(f(this.value))
            is Failure -> this
        }

    fun <U> mapFailure(f: (E) -> U): Outcome<U, T> =
        when (this){
            is Success -> this
            is Failure -> Failure(f(this.error))
        }

    fun <B> fold(ifEmpty: (E) -> B, ifNotEmpty: (T) -> B): B =
        when (this){
            is Success -> ifNotEmpty(this.value)
            is Failure -> ifEmpty(this.error)
        }


    companion object {
        fun <T: Any> tryThis(block: () -> T): Outcome<Throwable, T> =
            try {
                Success(block())
            } catch (e: Throwable){
                Failure(e)
            }
    }
}

data class Success<T: Any>(val value: T): Outcome<Nothing, T>()
data class Failure<E>(val error: E): Outcome<E, Nothing>()


inline fun <T: Any, U: Any, E> Outcome<E, T>.flatMap(f: (T) -> Outcome<E, U>): Outcome<E, U> =
    when (this) {
        is Success<T> -> f(value)
        is Failure<E> -> this
    }

inline fun <E, T: Any>Outcome<E, T>.mapNullableError(f: (T) -> E?): Outcome<E, Unit> =
    when (this){
        is Success<T> -> {
            val error = f(this.value)
            if (error == null ) Success(Unit) else Failure(error)
        }
        is Failure<E> -> this
    }

inline fun <T: Any, E> Outcome<E, T>.onFailure(block: (E) -> T): T = // acts like a recover from failure
    when (this) {
        is Success<T> -> value
        is Failure<E> -> block(error)
    }